app/Actions/AlbumAuthorisationProvider.php

```diff
@@ -0,0 +1,523 @@
+	 * {@link AlbumAuthorisationProvider::applyVisibilityFilter()} for a
+	 * specification of the rules when an album is visible.
+	 * In other cases, the following holds:
+	 *  - the root album is visible if and only if the user is authenticated
+	 *  - the built-in smart albums are visible, if
+	 *     - the user is authenticated and is granted the right of uploading, or
+	 *     - the album is the album of recent photos and public by configuration, or
+	 *     - the album is the album of starred photos and public by configuration
+	 *
+	 * Note, this method tries to minimize DB queries and any overhead due
+	 * to hydration of models.
+	 * If an actual instance of a {@link AbstractAlbum} model is passed in,
+	 * then the DB won't be queried at all, because all checks are performed
+	 * on the values of the already hydrated model.
+	 * If an ID is passed, then the method runs a very efficient COUNT
+	 * query on the DB.
+	 * In particular, no {@link Album} nor {@link TagAlbum} model is hydrated
+	 * to avoid any overhead.
+	 *
+	 * Tips for usage:
+	 *  - If you already have a {@link AbstractAlbum} instance, pass that.
+	 *    This is most efficient.
+	 *  - If you do not have a {@link AbstractAlbum} instance, but you will
+	 *    need one later anyway, then use {@link AlbumFactory} to first fetch
+	 *    the album from DB and pass the album.
+	 *    This avoids a second DB query later.
+	 *  - If you do not have a {@link AbstractAlbum} instance, and you won't
+	 *    need one later, simply pass the ID of the album.
+	 *    This avoids the overhead of model hydration.
+	 *
+	 * @param string|int|AbstractAlbum|null $albumModelOrID
+	 *
+	 * @return bool
+	 */
+	public function isVisible($albumModelOrID): bool
+	{
+		if (AccessControl::is_admin()) {
+			return true;
+		}
+
+		// deal with root album first
+		if (empty($albumModelOrID)) {
+			return AccessControl::is_logged_in();
+		}
+
+		/** @var ?AbstractAlbum $album */
+		/** @var int|string $albumID */
+		list($albumID, $album) = $this->disassembleAlbumParameter($albumModelOrID);
+
+		// Deal with built-in smart albums
+		if ($this->albumFactory->isBuiltInSmartAlbum($albumID)) {
+			return $this->isAuthorizedForSmartAlbum($albumID);
+		}
+
+		// Deal with albums that are actual models.
+		// If we already have an instance of a model, then avoid an unnecessary
+		// DB query.
+		// We perform the visibility checks directly on the album model.
+		// The semantics of these checks must be kept in sync with the
+		// checks in `applyVisibilityFilter`.
+		if ($album) {
+			/* @var BaseAlbum $album */
+
+			if (!AccessControl::is_logged_in()) {
+				return !$album->requires_link && $album->is_public;
+			} else {
+				$userID = AccessControl::id();
+
+				return
+					($album->owner_id === $userID) ||
+					(!$album->requires_link && $album->shared_with()->where('user_id', '=', $userID)->count()) ||
+					(!$album->requires_link && $album->is_public);
```

Comment:
It looks like `!$album->requires_link` is unnecessarily repeated? Could be written as:
```
(!$album->requires_link && ($album->shared_with()->where('user_id',` '=', $userID)->count() || $album->is_public))
```